Initial OrderStatisticMap/Set

lembergerth opened the pull request:
Provides the `OrderStatisticMap` and `OrderStatisticSet` interface
with naive implementations.

The interfaces provide creator methods that return objects of the implementing `Naive*` classes.
The `OrderStatistic*TestSuite` classes aim to provide a small test suite for the current and future `OrderStatistic*` implementations.

lembergerth committed:
Initial OrderStatisticMap/Set with OrderStatisticSet Tests

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticMap.java:
```diff
@@ -0,0 +1,138 @@
+import java.util.Map;
+import java.util.NavigableMap;
+
+public interface OrderStatisticMap<K, V> extends NavigableMap<K, V> {
```
Some JavaDoc here that explains what this is would be nice.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticMap.java:
```diff
@@ -0,0 +1,138 @@
+   * @throws IndexOutOfBoundsException if the given rank is out of the range of this set (i.e.,
+   *     pRank &lt; 0 || pRank &gt;= size)
+   */
+  K getKeyByRank(int pIndex);
```
Is this operation common enough to have its own method?
If yes, it could be a default method delegating to `getEntryByRank`, otherwise we could leave it out.

lembergerth replied:
I wanted to provide it in reference to `NavigableMap`.
It also includes methods `lower*`, `floor*`, etc. for `Entry` and `Key`.
But currently I'm not using or needing it.

PhilippWendler commented on src/org/sosy_lab/common/collect/NaiveOrderStatisticMap.java:
```diff
@@ -0,0 +1,162 @@
+ *
+ * <p>The class wraps a {@link NavigableMap} object and delegates all methods inherited from the
+ * <code>NavigableMap</code> interface to that. For the methods particular to the <code>
+ * OrderStatisticMap</code> interface, it provides naive implementations that do not guarantee any
```
Can we do better than "no guarantee"? At least something like O(n) should be possible, right? This should then also be stated on the public `create` methods of the interface.

PhilippWendler commented on src/org/sosy_lab/common/collect/NaiveOrderStatisticMap.java:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  /** Creates a new empty OrderStatisticMap using natural ordering. */
+  static <K, V> NaiveOrderStatisticMap<K, V> createMap() {
```
These non-public factory methods are not strictly necessary, the constructor could be made package-private. But if these factory methods help to reduce code, then keep them.

lembergerth replied:
I'm not sure whether it's really *less* code, but it circumvents the existence of
`OrderStatisticMap(Map)` and `OrderStatisticMap(NavigableMap)`, and thus is more concrete and less error-prone.

PhilippWendler commented on src/org/sosy_lab/common/collect/NaiveOrderStatisticMap.java:
```diff
@@ -0,0 +1,162 @@
+
+  @Override
+  public K getKeyByRank(int pIndex) {
+    return NaiveOrderStatisticSet.createSetWithSameOrder(delegate.navigableKeySet())
```
Doesn't this copy the whole set on every access? This seems like a really expensive operation, and most callers of getters will assume that they do not involve large memory allocations. Is there no easy way to avoid this?

lembergerth committed:
OrderStatistic classes have improved Javadoc

lembergerth committed:
Improve OrderStatisticMapTestSuite and fix bug

lembergerth committed:
NaiveOrderStatisticMap handles #rankOf and #getKeyByRank more efficiently.

Does not create new NaiveOrderStatisticSet on every #rankOf and #getKeyByRank
call.

lembergerth committed:
Fix bugs in NaiveOrderStatisticMap.

Properly use delegation in NaiveOrderStatisticMap,
only use view of NaiveOrderStatisticSet for keySet view,
don't create a new set that doesn't reflect changes in map.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticMapTestSuite.java:
```diff
@@ -0,0 +1,369 @@
+    OrderStatisticMap<String, String> l1 = createMap();
+    @Var OrderStatisticMap<String, String> l2 = createMap();
+
+    // Check that the map sorts its elements
```
`equals` of maps does not depend on ordering. I recommend `EqualsTester` for testing `equals`, because one should also check for example that such a map is equal to a `HashMap` with the same contents etc. But tests that the `NavigableMapTestSuiteBuilder` already does of course need not be written manually.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticMapTestSuite.java:
```diff
@@ -0,0 +1,369 @@
+    Entry<String, String> toAdd = ELEMS_BELOW.get(2);
+
+    putEntry(subMap, toAdd);
+    Assert.assertTrue(containsEntry(subMap, toAdd));
```
`assertTrue` and `assertFalse` create hard-to-debug test cases and should be avoided. The various Truth methods are much better.

For consistency, it might make sense to switch to Truth for all assertions in these classes.

Furthermore, these methods can be statically imported.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticMapTestSuite.java:
```diff
@@ -0,0 +1,369 @@
+    Assert.assertFalse(containsEntry(subMap, toAdd));
+  }
+
+  @Test(expected = IllegalArgumentException.class)
```
`expected` is bad except in single-statement test cases. Please use `try { ...; fail(); } catch (IllegalArgumentException expected) { }` with as little code inside the `try` as possible.

stieglma replied:
why would you use try/fail instead of an expected excpeption rule? In my opinion this creates much nicer, and especially more readable code. Moreover the nested try-catches should be separate unit-tests, should they?

PhilippWendler replied:
`ExpectedException` has a single advantage over `Test(expected=)`: it lets you reduce the scope where the exception is expected from the whole method to a range of code between the call to ExpectedException and the end of the method. So, it is better than `Test(expected=)`, and if the call to ExpectedException is the second-to-last statement, then it works fine.

However, both methods have one serious problem: If the expected exception is not thrown in the last statement of the method but before, the remainder of the test case will be silently ignored. This is error-prone, and it seems [Google has found out](https://github.com/google/error-prone/issues/219#issuecomment-71244204) that many developers assume that ExpectedException does some magic and actually manages to resume code execution after the expected exception is thrown. So my understanding from following the development of Guava, Truth, and Error-Prone is that both `ExpectedException` and `Test(expected=)` are [not](https://github.com/google/error-prone/issues/219#issuecomment-86110692) [worth](http://errorprone.info/bugpattern/TestExceptionChecker) it in most cases and the old `try-fail-catch` way is [recommended](https://stackoverflow.com/a/38866403/396730) (Christian Gruber is one of the Google developers of these projects) because it is less error-prone *and* allows you to safely write code after the expected exception. Error-Prone also has a check for not forgettting the `fail()` in the `try`, which would be the most important kind of error in this style of writing tests.

All this is true only until we have a newer JUnit that supports writing something like `assertThrows()` with a lambda. I guess then this should be preferred.

Now, independently from this it is of course good if tests are as small as possible, and this could also be improved in this class by splitting them into more methods. And if you follow that strictly, you might decide that the remaining risk of using `ExpectedException` is quite small and that it is fine to use. That would be fine.
For the current set of tests, however, I recommended using `try-fail-catch` because the exception handling is the more important problem than the somewhat large tests (the latter are only a problem for debugging but not necessarily error-prone), and because these tests already use `try-catch` in addition to `Test(expected=)`, switching completely to `try-fail-catch` would be easier.

stieglma replied:
wow, thanks for the detailed response! I never thought about it that way. We never took the actual location where the exception is thrown into account (our unit-tests follow the arrange-act-assert principle and the the to-be-thrown exception is declared in the arrange block). What we do always is check the error message (and by "contract" if any method we implemented throws an exception, it has to have a proper message), and thus we still can be sure that we caught the right exception

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticSetTestSuite.java:
```diff
@@ -0,0 +1,282 @@
+    OrderStatisticSet<String> set = createSet(ELEMS);
+    NavigableSet<String> subSet =
+        set.subSet(ELEMS[1], /* fromInclusive= */ true, ELEMS[3], /*
+    toInclusive=
```
please remove the line breaks in this comment

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticMapTestSuite.java:
```diff
@@ -0,0 +1,369 @@
+
+public abstract class OrderStatisticMapTestSuite {
+
+  @SuppressWarnings("unchecked")
```
This seems unnecessary, and all other occurrences in this file, too.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticMap.java:
```diff
@@ -0,0 +1,173 @@
+   *     as the given map
+   */
+  static <K, V> OrderStatisticMap<K, V> createWithSameOrder(
+      NavigableMap<K, ? extends V> pNavigableMap) {
```
This parameter could accept a `SortedMap`, right?
(same for `OrderStatisticSet`)

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticSet.java:
```diff
@@ -0,0 +1,150 @@
+   * natural ordering. The returned map guarantees performance only in O(n) for the operations
+   * specific to the OrderStatisticSet interface.
+   */
+  static <E> OrderStatisticSet<E> createWithNaturalOrder(Collection<E> pCollection) {
```
We could accept an `Iterable` here. By using `Iterables.addAll`, we do not need more code and get the same performance benefit as before if a `TreeMap` is passed.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticSet.java:
```diff
@@ -0,0 +1,150 @@
+  OrderStatisticSet<E> descendingSet();
+
+  @Override
+  OrderStatisticSet<E> subSet(
```
Btw., what is the rank of elements in a subset? Do the elements keep the same rank as in the original set, or do they get a new rank according to the contents of the subset? I would strongly prefer the latter. Are there tests for this?

The same question of course exists for all other subset views returned by some methods, and for the map methods. Furthermore, how does it work on the `descendingSet`?

The answers to these questions should be documented in the JavaDoc, either globally for all view-returning methods, or per method.

lembergerth replied:
I wouldn't even have thought of the first possibility.

`rankOf` and the other rank-methods work on the rank of the element in the `OrderStatistic*` collection the method is called on.
So if i call `subSet.rankOf(obj)`, it was obvious to me that, as the JavaDoc states, it returns the rank of `obj` in that specific set, i.e., `subSet`, not some backing set that the user may not even know of.

I have added clarifying Javadocs to the interface methods.

PhilippWendler reviewed (changes requested):
Please also run "ant format-source", it did a little bit of JavaDoc formatting.

lembergerth committed:
OrderStatisticMap#getKeyByRank has default implementation

lembergerth committed:
OrderStatistic*#createWithSameOrder creator methods accept Sorted classes, not just Navigable ones

lembergerth committed:
Format source

lembergerth committed:
Remove unnecessary SupressWarnings("unchecked")

lembergerth committed:
OrderStatistic* test suites use EqualsTester for comparison of collections

lembergerth committed:
OrderStatistic* test suites improved: Use Guava Truth for test assertions, add tests for #rankOf* methods

lembergerth committed:
OrderStatisticSet#createWithNaturalOrder accepts more generic Iterable, not just Collection

lembergerth committed:
Javadoc of OrderStatistic* interface describes behavior of element/key rank for views

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticMapTestSuite.java:
```diff
@@ -145,43 +144,50 @@ public void testSubmapView_mutation() {
-    } catch (IllegalArgumentException e1) {
+      fail();
+    } catch (IllegalArgumentException expected) {
       try {
```
Nitpick: I wouldn't nest these, but just have consecutive `try-catch` statements.

lembergerth replied:
No, of course - this is a left-over of the `@Test(expected=..)` annotations

lembergerth committed:
Try-catch blocks in OrderStatistic* test suites not nested
